Slim/Schema/Contributor.pm

```diff
@@ -89,6 +89,20 @@ sub defaultContributorRoles {
 	return grep { __PACKAGE__->typeToRole($_) < MIN_CUSTOM_ROLE_ID } contributorRoles();
 }
 
+sub splitDefaultAndCustomRoles {
+	my $self = shift;
+	my $roles = shift;
+
+	return {
+		defaultRoles => (join(',', grep {
+			my $x = $_; grep {$x eq $_} defaultContributorRoles()
+			} split(',', $roles || ''))),
+		userDefinedRoles => (join(',', grep {
+			my $x = $_; !grep {$x eq $_} defaultContributorRoles()
+			} split(',', $roles || ''))),
+	};
```

Comment:
I'd rewrite this to have the function `return ($defaultRoles, $userDefinedRoles)`. That way you don't have to deal with a hash ref in the caller.

Plus it's pretty inefficient, as the `$roles` are split and processed twice, and `defaultContributorRoles()` is called twice for each element of the `$roles` list. Not that that would really matter in this case. But I'd still
* define the two variables for the result
* store the result of `defaultContributroRoles()` in another variable
* loop over split `$roles` _once_ only
* assign to each of those variables where applicable
* return the `join`ed version of the two lists

Should be easier to read and handle.